src/viur/core/modules/moduleconf.py

```diff
@@ -68,16 +68,36 @@ def get_by_module_name(cls, module_name: str) -> None | SkeletonInstance:
 @StartupTask
 def read_all_modules():
     db_module_names = [m["name"] for m in db.Query("viur-module-conf").run(999)]
-    module_names = dir(conf.main_app.vi)
-
-    for module_name in module_names:
-        module = getattr(conf.main_app.vi, module_name)
-        if isinstance(module, Module):
+    visited_modules = set()
+
+    def collect_modules(parent, depth: int = 0, prefix: str = "") -> None:
+        """Recursively collects all routable modules for the vi renderer"""
+        if depth > 10:
+            logging.warning(f"Reached maximum recursion limit of {depth} at {parent=}")
+            return
+
+        for module_name in dir(parent):
+            module = getattr(parent, module_name, None)
+            if not isinstance(module, Module):
+                continue
+            if module in visited_modules:
+                # Some modules reference other modules as parents, this will
+                # lead to infinite recursion. We can avoid reaching the
+                # maximum recursion limit by remembering already seen modules.
+                if conf.debug.trace:
+                    logging.debug(f"Already visited and added {module=}")
+                continue
+            visited_modules.add(module)
             if module_name not in db_module_names:
                 skel = conf.main_app._moduleconf.addSkel()
-                skel["key"] = db.Key("viur-module-conf", module_name)
-                skel["name"] = module_name
+                skel["key"] = db.Key("viur-module-conf", f"{prefix}{module_name}")
```

Comment:
Make this less static!
```suggestion
                skel["key"] = db.Key(skel.kindName, f"{prefix}{module_name}")
```